Integrates H2O Driverless AI MOJO & Python Scoring Pipeline into Apache MiNiFi C++

james94 opened the pull request:
This pull request resolves GitHub filed issue: [Add Python Wrapper MOJO Deploy Ex to Apache MiNiFiCPP Python](https://github.com/h2oai/dai-deployment-examples/issues/19) #19 

Short Summary:

I have developed deployment examples for executing the Driverless AI MOJO Scoring Pipeline and Python Scoring Pipeline within a MiNiFi C++ Data Flow. In **mojo-py-minificpp** folder, I provide walkthroughs for both deployment examples. Additionally, I also provide video walkthrough as an alternative to a written walkthrough. This contribution is a Cloudera Integration Point for Cloudera Data Flow (CDF) and Driverless AI that makes it easier to do model deployment and model scoring on the edge.

Long Summary:

**MiNiFi C++ and H2O Driverless AI Py Wrapper MOJO Integration** via Custom Python Processors:

Integrates MiNiFi C++ with H2O Driverless AI by using Driverless AI's MOJO Scoring Pipeline (MSP) in the C++ Runtime with the Python Wrapper and MiNiFi's Custom Python Processor. Uses only one Python Processor to execute the MOJO Scoring Pipeline to do batch scoring or real-time scoring for one or more predicted labels on tabular test data in the incoming flow file content. If the tabular data is one row, then the MOJO does real-time scoring. If the tabular data is multiple rows, then the MOJO does batch scoring. I would like to contribute the **mojo-py-minificpp** as a new deployment example, which could be transitioned to a production ready deployment template.

**1 custom python processor** created for MiNiFi and MOJO:

**ExecuteDaiMojoScoringPipeline** - Executes H2O Driverless AI's MOJO Scoring Pipeline in C++ Runtime Python Wrapper to do batch scoring or real-time scoring on a frame of data within each incoming flow file. 

**MiNiFi C++ and H2O Driverless AI Python Scoring Pipeline Integration** via Custom Python Processors:

Integrates MiNiFi C++ with H2O's Driverless AI by Using Driverless AI's Python Scoring Pipeline (PSP) and MiNiFi's Custom Python Processors. Uses the Python Processors to execute the Python Scoring Pipeline scorer to do batch scoring and real-time scoring for one or more predicted labels on test data in the incoming flow file content. 

**2 custom python processors** created for MiNiFi and Python Scoring Pipeline:

**ExecuteDaiPspRealTimeScoring** - Executes H2O Driverless AI's Python Scoring Pipeline to do interactive scoring (real-time) scoring on an individual row or list of test data within each incoming flow file.

**ExecuteDaiPspBatchScoring** - Executes H2O Driverless AI's Python Scoring Pipeline to do batch scoring on a frame of data within each incoming flow file.

**1 custom python processor** created for MiNiFi and Datatable:

**ConvertDsToCsv** - Converts data source of incoming flow file to csv.

nkpng2k commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,160 @@
+
+## Prerequisites
+
+- Java >= 7 runtime (JDK >= 1.7)
```
ummm C++ / python? so why is there java dependencies.

james94 replied:
Hey @nkpng2k, here is the link to the appropriate README for MiNiFi C++ Py Driverless AI MOJO: https://github.com/james94/dai-deployment-examples/tree/mojo-py-minificpp/mojo-py-minificpp

nkpng2k commented on mojo-nifi/README.md:
```diff
@@ -0,0 +1,160 @@
+
+### Process For Developing the Processor
+
+For developing the NiFi custom processor, I used Maven and Eclipse IDE. When using NiFi with Maven, there is the option to build a starter template for the custom processor you are going to build. To generate the starter template, I used the command: `mvn archetype:generate`. Next my plan was to work on the processor's code in Eclipse. I ran the command `mvn install` to compile the project. Then  ...
```
see comments from other pr.

james94 replied:
@nkpng2k you are looking at the README inside the mojo-nifi folder. Check out the "Process for Developing the MiNiFi C++ Py Processor" section in the appropriate README inside the mojo-py-minificpp folder: https://github.com/james94/dai-deployment-examples/tree/mojo-py-minificpp/mojo-py-minificpp

nkpng2k commented on mojo-nifi/model-deployment/apps/nifi/nifi-nar-bundles/nifi-h2o-record-bundle/nifi-h2o-record-processors/src/main/resources/META-INF/services/org.apache.nifi.processor.Processor:
```diff
@@ -0,0 +1,15 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+org.apache.nifi.processors.h2o.record.ExecuteDaiMojoScoringPipeline
```
nit: new line at end of files.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/conf/config-msp-interactive-scoring.yml:
```diff
@@ -0,0 +1,76 @@
+      destination name: PutFile
+
+Remote Processing Groups: []
+Provenance Reporting:
```
nit: new line please.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/conf/config-msp-batch-scoring.yml:
```diff
@@ -0,0 +1,76 @@
+      destination name: PutFile
+
+Remote Processing Groups: []
+Provenance Reporting:
```
nit: new line.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/modules/DaiPythonScorer.py:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env python
+
+from scoring_h2oai_experiment_0abac7c4_8993_11ea_ae54_0242ac110002 import Scorer
```
nit: new line.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py:
```diff
@@ -0,0 +1,119 @@
+            ff_attr_name = pred_header_list[i] + "_pred"
+            flow_file.addAttribute(ff_attr_name, str(preds_list[i]))
+            log.info("getAttribute({}): {}".format(ff_attr_name, flow_file.getAttribute(ff_attr_name)))
+        session.transfer(flow_file, REL_SUCCESS)
```
nit: newline.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py:
```diff
@@ -0,0 +1,119 @@
+import pandas as pd
+import datatable as dt
+sys.path.append(os.environ['HOME'] + "/nifi-minifi-cpp-0.7.0/modules/")
+from DaiPythonScorer import *
```
so this seems to be using dai python scoring pipeline. I would actually advise to use the dai mojo cpp that way you can simply use the mojo... reduces need for complex python environment dependencies associated with python scoring pipeline.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py:
```diff
@@ -0,0 +1,119 @@
+        pred_header_str = context.getProperty("Predicted Label(s)")
+        pred_header_list = pred_header_str.split(",")
+        # load tabular data str into datatable, convert to pd df, then to list of lists
+        test_dt_frame = dt.Frame(read_cb.content)
```
why read to dt if never using dt for anything? `pd.read_csv`

james94 replied:
@nkpng2k I read the string content to the dt because dt is able to properly load the string into the table whereas pandas and most other libraries are not able to properly load the data into their table.

james94 commented:
Hey @nkpng2k. Before I start addressing your comments. Is there a way I can make these two pull requests not overlap. For instance, I created a new branch off the master branch, then I started by making the NiFi pull request #18. Then I created a new branch off the master branch for MiNiFi C++ Py, but I noticed the NiFi updates came into the MiNiFi C++ Py branch. Can I make these branches independent?

nkpng2k commented:
@james94 I'm looking at your commits and it seems like there are like 40 different commits... now if they are not mixed together. E.g. you know which commits relate to the python as opposed to the java, then maybe you can cherry-pick out the commits to a separated branch. https://stackoverflow.com/questions/1670970/how-to-cherry-pick-multiple-commits, might be a good idea to investigate this/something similar.

so if you were to follow the link, create new branch cherry-pick out the commits related to specifically the python, close the branch/pr with mixed commits, open new pr with just python. (make sure to checkout from top of master for new branch)

james94 commented:
@nkpng2k I am going to close this PR, then create a new branch based on top of master. Then open a new PR with just Python, MiNiFi and Driverless AI MOJO.
